compiler/src/org.graalvm.compiler.phases.common/src/org/graalvm/compiler/phases/common/ReassociationPhase.java

```diff
@@ -0,0 +1,121 @@
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.graalvm.compiler.phases.common;
+
+import org.graalvm.compiler.debug.DebugContext;
+import org.graalvm.compiler.graph.NodeBitMap;
+import org.graalvm.compiler.loop.LoopEx;
+import org.graalvm.compiler.loop.LoopsData;
+import org.graalvm.compiler.nodes.NodeView;
+import org.graalvm.compiler.nodes.StructuredGraph;
+import org.graalvm.compiler.nodes.ValueNode;
+import org.graalvm.compiler.nodes.calc.BinaryArithmeticNode;
+import org.graalvm.compiler.nodes.util.GraphUtil;
+import org.graalvm.compiler.phases.Phase;
+
+/**
+ * Rearrange {@link BinaryArithmeticNode#isAssociative() associative binary operations} for loop
+ * invariants and constants.
+ */
+public class ReassociationPhase extends Phase {
+
+    @SuppressWarnings("try")
+    @Override
+    protected void run(StructuredGraph graph) {
+        DebugContext debug = graph.getDebug();
+        try (DebugContext.Scope s = debug.scope("ReassociationPhase")) {
+            reassociateInvariant(graph, debug);
+            reassociateConstant(graph, debug);
+        } catch (Throwable e) {
+            throw debug.handle(e);
+        }
+        graph.setAfterReassociation();
+    }
+
+    /**
+     * Re-associate loop invariant so that invariant parts of the expression can move outside of the
+     * loop.
+     */
+    @SuppressWarnings("try")
+    private static void reassociateInvariant(StructuredGraph graph, DebugContext debug) {
+        LoopsData loopsData = new LoopsData(graph);
+        int iterations = 0;
+        try (DebugContext.Scope s = debug.scope("ReassociateInvariants")) {
+            boolean changed = true;
+            while (changed) {
+                changed = false;
+                for (LoopEx loop : loopsData.loops()) {
+                    changed |= loop.reassociateInvariants();
+                }
+                loopsData.deleteUnusedNodes();
+                iterations++;
+                debug.dump(DebugContext.VERBOSE_LEVEL, graph, "after iteration %d", iterations);
```

Comment:
pls make that a VERY_DETAILED level and add the string "reassociation:" to the message

Reply:
Since it has the "ReassociateInvariants" debug scope, is it really needed to add the "reassociation" to the message?